### PR TITLE
[Refunds] Disable refund when order is free

### PR DESCRIPTION
**Pull request by AdamKasp**
Screenshot 2020-01-30 at 13 08 10
Screenshot 2020-01-30 at 13 08 24

### CODE REVIEW

File: src/Action/Admin/OrderRefundsListAction.php [hunk 1]
```diff
@@ -57,7 +57,11 @@ public function __invoke(Request $request): Response
         $order = $this->orderRepository->findOneByNumber($request->attributes->get('orderNumber'));
 
         if (!$this->orderRefundsListAvailabilityChecker->__invoke($request->attributes->get('orderNumber'))) {
-            return $this->redirectToReferer($order);
+            if ($order->getTotal() === 0) {
+                return $this->redirectToReferer($order, 'sylius_refund.free_order_should_not_be_refund');
+            }
+            return $this->redirectToReferer($order, 'sylius_refund.order_should_be_paid');
+
```
Comment (Tomanhez) [inline src/Action/Admin/OrderRefundsListAction.php]: redundant blank line

File: features/preventing_from_refund_free_order.feature [hunk 2]
```diff
@@ -0,0 +1,28 @@
+@refunds
+Feature: Preventing from refund a free order
```
Comment (GSadee) [inline features/preventing_from_refund_free_order.feature]:
```suggestion
Feature: Being unable to refund a free order
```

File: features/preventing_from_refund_free_order.feature [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    In order to give back money to the customer only if it is possible
+    As an Administrator
+    I want to not be able to generate refund for free order
```
Comment (GSadee) [inline features/preventing_from_refund_free_order.feature]:
```suggestion
    I want not to be able to generate a refund for a free order
```

File: features/preventing_from_refund_free_order.feature [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        And the store allows paying with "Space money"
+        And there is a customer "Lucifer Morningstar" that placed an order "#0000001"
+        And the customer bought 3 "Witcher Sword" products
+        And the customer "Lucifer Morningstar" addressed it to "Seaside Fwy", "90802" "Los Angeles" in the "United States" with identical billing address
+        And the customer chose "Free" shipping method
+        And this order is already paid
+        And I am logged in as an administrator
+
+        @ui
+        Scenario: Not being able to refund free order
```
Comment (GSadee) [inline features/preventing_from_refund_free_order.feature]:
```suggestion
        Scenario: Being unable to refund a free order
```

File: features/preventing_from_refund_free_order.feature [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            When I view the summary of the order "#0000001"
+            Then I should not be able to see refunds button
```
Comment (GSadee) [inline features/preventing_from_refund_free_order.feature]:
```suggestion
            Then I should not see refunds button
```
or
```suggestion
            Then I should not be able to refund an order
```

File: features/preventing_from_refund_free_order.feature [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+        @ui
+        Scenario: Not being able to open refund page when order is free
+            When I try to refund some units of order "#0000001"
+            Then I should be redirected to the order "#0000001" show page
+            And I should be notified that I cannot refund free order
```
Comment (GSadee) [inline features/preventing_from_refund_free_order.feature]:
```suggestion
            And I should be notified that I cannot refund a free order
```

File: src/Action/Admin/OrderRefundsListAction.php [hunk 7]
```diff
@@ -57,7 +57,11 @@ public function __invoke(Request $request): Response
         $order = $this->orderRepository->findOneByNumber($request->attributes->get('orderNumber'));
 
         if (!$this->orderRefundsListAvailabilityChecker->__invoke($request->attributes->get('orderNumber'))) {
-            return $this->redirectToReferer($order);
+            if ($order->getTotal() === 0) {
+                return $this->redirectToReferer($order, 'sylius_refund.free_order_should_not_be_refund');
+            }
+            return $this->redirectToReferer($order, 'sylius_refund.order_should_be_paid');
```
Comment (GSadee) [inline src/Action/Admin/OrderRefundsListAction.php]: Missing blank line before `return`

File: src/Resources/translations/flashes.en.yml [hunk 8]
```diff
@@ -1,5 +1,7 @@
 sylius_refund:
     at_least_one_unit_should_be_selected_to_refund: 'At least one unit should be selected to refund'
+    error_occurred: 'Unexpected error occurred'
+    free_order_should_not_be_refund: 'You can not refund free order'
```
Comment (GSadee) [inline src/Resources/translations/flashes.en.yml]:
```suggestion
    free_order_should_not_be_refund: 'You cannot refund a free order'
```

File: tests/Behat/Context/Ui/ManagingOrdersContext.php [hunk 9]
```diff
@@ -106,4 +106,23 @@ public function thisOrderSPaymentStateShouldBe(OrderInterface $order, string $or
             'paymentState' => $orderPaymentState,
         ]));
     }
+
+    /**
+     * @Then I should be redirected to the order :order show page
+     */
+    public function iShouldBeRedirectedToTheOrderShowPage(OrderInterface $order): void
+    {
+        $this->showPage->isOpen(['id' => $order->getId()]);
```
Comment (GSadee) [inline tests/Behat/Context/Ui/ManagingOrdersContext.php]: Without `Assert::true()` this step does nothing

File: features/being_unable_to_refund_free_order.feature [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        And this order is already paid
+        And I am logged in as an administrator
+
+        @ui
+        Scenario: Being unable to refund a free order
+            When I view the summary of the order "#0000001"
+            Then I should not see refunds button
+
+        @ui
+        Scenario: Not being able to open refund page when order is free
```
Comment (GSadee) [inline features/being_unable_to_refund_free_order.feature]: Should look similar to the scenario above

File: src/Checker/OrderRefundingAvailabilityChecker.php [hunk 11]
```diff
@@ -27,7 +27,8 @@ public function __invoke(string $orderNumber): bool
 
         return in_array(
             $order->getPaymentState(),
-            [OrderPaymentStates::STATE_PAID, OrderPaymentStates::STATE_PARTIALLY_REFUNDED]
-        );
+            [OrderPaymentStates::STATE_PAID, OrderPaymentStates::STATE_PARTIALLY_REFUNDED]) &&
```
Comment (lchrusciel) [inline src/Checker/OrderRefundingAvailabilityChecker.php]: Something went wrong with the formatting
```suggestion
            [OrderPaymentStates::STATE_PAID, OrderPaymentStates::STATE_PARTIALLY_REFUNDED] &&
```

Whole condition could look like this:
```php
        return in_array($order->getPaymentState(), [
            OrderPaymentStates::STATE_PAID, 
            OrderPaymentStates::STATE_PARTIALLY_REFUNDED,
        ] ) && $order->getTotal() !== 0;
```
or something similar

File: src/Checker/OrderRefundsListAvailabilityChecker.php [hunk 12]
```diff
@@ -27,7 +27,8 @@ public function __invoke(string $orderNumber): bool
 
         return in_array(
             $order->getPaymentState(),
-            [OrderPaymentStates::STATE_PAID, OrderPaymentStates::STATE_PARTIALLY_REFUNDED, OrderPaymentStates::STATE_REFUNDED]
-        );
+            [OrderPaymentStates::STATE_PAID, OrderPaymentStates::STATE_PARTIALLY_REFUNDED, OrderPaymentStates::STATE_REFUNDED]) &&
```
Comment (lchrusciel) [inline src/Checker/OrderRefundsListAvailabilityChecker.php]: Same as above

**Comment by AdamKasp** [comment]
> I'm still wondering if we cannot put blocking logic somewhere to state machine with guards.

Good idea for some refactor 💯

**Comment by GSadee** [comment]
Thank you, Adam! :tada:

**Pull request merged**